src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -675,12 +679,14 @@ void ShenandoahOldGeneration::handle_failed_promotion(Thread* thread, size_t siz
   static size_t epoch_report_count = 0;
   auto heap = ShenandoahGenerationalHeap::heap();
 
-  size_t promotion_reserve;
-  size_t promotion_expended;
-
   const size_t gc_id = heap->control_thread()->get_gc_id();
 
+  AtomicAccess::inc(&_promotion_failure_count);
```

Comment:
Just noticing that in the next code block, we acquire the heap->lock().  Could we just use that same heap lock to protect adjustments to _promotion_failure_count and _promotion_failure_words and then we would need to use Atomic access operations?

Reply:
Should probably be ok to read these variables without lock.  We're only logging these results when evacuation is no longer happening.  Right?

Reply:
Now that we have this new log message, can we get rid of the "promotion failure messages" for individual objects?

Reply:
If we made this a `log_debug(gc, plab)` message, we could make the whole block of code conditional on the log level being enabled.